core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/batchutilities/BatchManager.java

```diff
@@ -0,0 +1,212 @@
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicInteger;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.utils.ThreadFactoryBuilder;
+
+/**
+ * Implementation of a generic buffer for automatic request batching.
+ * @param <T> the type of an outgoing request.
+ * @param <U> the type of an outgoing response.
+ * @param <V> the type of an outgoing batch response.
+ */
+@SdkInternalApi
+public class BatchManager<T, U, V> {
+    // Just a number from the cloudwatch metric publisher for now. Not sure if I should choose a different max task queue size?
+    private static final int MAXIMUM_TASK_QUEUE_SIZE = 128;
+
+    private final BatchingMap<T> batchGroupIdToIdToRequest;
+    private final BatchingMap<CompletableFuture<U>> batchGroupIdToIdToResponse;
+    private final Map<String, ScheduledFlush> scheduledFlushTasks;
+    private final Map<String, AtomicInteger> currentIds;
+    private final BatchAndSendFunction<T, V> batchingFunction;
+    private final UnpackBatchResponseFunction<V, U> unpackResponseFunction;
+    private final ScheduledExecutorService scheduledExecutor;
+    private final ThreadFactory threadFactory;
+    private final ExecutorService executor;
+    private final Duration maxBatchOpenInMs;
+    private final int maxBatchItems;
+
+    public BatchManager(int maxBatchItems, Duration maxBatchOpenInMs,
+                       BatchAndSendFunction<T, V> batchingFunction,
+                       UnpackBatchResponseFunction<V, U> unpackResponseFunction) {
+        this.batchGroupIdToIdToRequest = new BatchingMap<>();
+        this.batchGroupIdToIdToResponse = new BatchingMap<>();
+        this.scheduledFlushTasks = new ConcurrentHashMap<>();
+        this.currentIds = new ConcurrentHashMap<>();
+        this.maxBatchItems = maxBatchItems;
+        this.maxBatchOpenInMs = maxBatchOpenInMs;
+        this.batchingFunction = batchingFunction;
+        this.unpackResponseFunction = unpackResponseFunction;
+        this.threadFactory = new ThreadFactoryBuilder().threadNamePrefix("batch-buffer").build();
+        this.scheduledExecutor = Executors.newSingleThreadScheduledExecutor(threadFactory);
+        this.executor = new ThreadPoolExecutor(1, 1, 0L, TimeUnit.MILLISECONDS,
```

Comment:
I was following the approach Matt used in the `CloudwatchMetricPublisher` where the `scheduledExecutor` just schedules tasks to run on the executor thread to ensure that there is only a single consumer and it is thread safe. When I was testing it with just a `scheduledExecutor` it would sometimes create new threads so I don't think it was the most thread safe. https://github.com/aws/aws-sdk-java-v2/blob/master/metric-publishers/cloudwatch-metric-publisher/src/main/java/software/amazon/awssdk/metrics/publishers/cloudwatch/CloudWatchMetricPublisher.java#L216

Reply:
I see. Can we add javadocs on the members just like how `executor` and `scheduledExecutor` are documented in `CloudwatchMetricPublisher`?